File: testing/testsrc/relu.py
import numpy as np

class relu:
  def forward(self, input):
    output = np.maximum(0, input)
    return output

  def backward(self, input):
    equation = np.vectorize(self.equationforderivative, otypes=[float])
    return equation(input)
    
  def equationforderivative(self, input):
    if input < 0:
      return 0
    elif input == 0:
      return None
    elif input > 0:    
      return 1

File: efficiency-testing/aiinpycpp/relu.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class relu {
  public: 
    vector <double> forward(vector <double> input) {
      transform(input.begin(), input.end(), input.begin(), [](double input){ return (0 < input) ? input : 0; });
      return input;
    }

    vector <double> backward(vector <double> input) {
      transform(input.begin(), input.end(), input.begin(), [](double input){ return (input <= 0) ? 0 : 1; });
      return input;
    }
};
